[Cache] Add url decoding of password in `RedisTrait` DSN

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #52420
| License       | MIT

This brings support for `@` char in passwords

nicolas-grekas commented:
Why not as a bug fix? This looks like one to me. For 5.4 then.

nicolas-grekas commented on src/Symfony/Component/Cache/Tests/Traits/RedisTraitTest.php:
```diff
@@ -14,15 +14,11 @@
-    public static function setUpBeforeClass(): void
-    {
-        if (!getenv('REDIS_CLUSTER_HOSTS')) {
-            self::markTestSkipped('REDIS_CLUSTER_HOSTS env var is not defined.');
```
but we still need the env var to run all the other test cases, don't we?

alexandre-daubois replied:
Yes, but this env var presence is already checked in `testCreateConnection` so it can be removed from here

nicolas-grekas commented on src/Symfony/Component/Cache/Traits/RedisTrait.php:
```diff
@@ -369,9 +369,9 @@ public static function createConnection(#[\SensitiveParameter] string $dsn, arra
+                    $params['parameters']['password'] = urldecode($auth[1]);
                 } else {
-                    $params['parameters']['password'] = $auth;
+                    $params['parameters']['password'] = urldecode($auth);
```
everywhere this?
```suggestion
                    $params['parameters']['password'] = rawurldecode($auth);
```

alexandre-daubois replied:
Good idea!

nicolas-grekas commented on src/Symfony/Component/Cache/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+7.1
+---
+
+ * Add url decoding of password in `RedisTrait` DSN
```
not needed for bug fixes

alexandre-daubois replied:
Oh alright, I thought it was considered as a feature. I'll update this as well as the base branch

alexandre-daubois committed:
[Cache] Add url decoding of password in `RedisTrait` DSN

alexandre-daubois commented:
I addressed your comments and rebased on 5.4. The fix is different from other branches, I'll provide a fix for those if needed :slightly_smiling_face:

nicolas-grekas commented:
Thank you @alexandre-daubois.

nicolas-grekas commented:
> I addressed your comments and rebased on 5.4. The fix is different from other branches, I'll provide a fix for those if needed 🙂

No need, that was easy to fix as part of the mege conflict.

Note that I didn't change CouchbaseCollectionAdapter nor CouchbaseBucketAdapter because while they have the same issue, they're more severely broken and would deserve a dedicated PR, if anyone is up to have a look ;)
